tests/testthat/test-model-compile-user_header.R

```diff
@@ -0,0 +1,104 @@
+
+file_that_exists <- 'placeholder_exists'
+file_that_doesnt_exist <- 'placeholder_doesnt_exist'
+file.create(file_that_exists)
+on.exit(if(file.exists(file_that_exists)) file.remove(file_that_exists), add=TRUE, after=FALSE)
+
+make_local_orig <- cmdstan_make_local()
+cmdstan_make_local(cpp_options = list("PRECOMPILED_HEADERS"="false"))
+on.exit(cmdstan_make_local(cpp_options = make_local_orig, append = FALSE), add = TRUE, after = FALSE)
+
+test_that("cmdstan_model works with user_header with mock", {
+  skip_if(os_is_macos())
+  tmpfile <- tempfile(fileext = ".hpp")
+  hpp <-
+  "
+  #include <stan/math.hpp>
+  #include <boost/math/tools/promotion.hpp>
+  #include <ostream>
+
+  namespace bernoulli_external_model_namespace
+  {
+      template <typename T0__,
+            stan::require_all_t<stan::is_stan_scalar<T0__>>* = nullptr>
+      inline typename boost::math::tools::promote_args<T0__>::type make_odds(const T0__ &
+                                                                                 theta,
+                                                                             std::ostream *pstream__)
+      {
+          return theta / (1 - theta);
+      }
+  }"
+  cat(hpp, file = tmpfile, sep = "\n")
+
+  with_mocked_cli(compile_ret = list(status = 0), info_ret = list(), code = expect_mock_compile(
+    expect_warning(
+      expect_no_warning({
+        mod <- cmdstan_model(
+          stan_file = testing_stan_file("bernoulli_external"),
+          exe_file = file_that_exists,
+          user_header = tmpfile
+        )
+      }, message = 'Recompiling is recommended'), # this warning should not occur because recompile happens automatically
+      'Retrieving exe_file info failed' # this warning should occur
+    )
+  ))
+
+  with_mocked_cli(compile_ret = list(status = 0), info_ret = list(), code = expect_mock_compile({
+    mod_2 <- cmdstan_model(
+      stan_file = testing_stan_file("bernoulli_external"),
+      exe_file = file_that_doesnt_exist,
+      cpp_options=list(USER_HEADER=tmpfile),
+      stanc_options = list("allow-undefined")
+    )
+  }))
+
+  # Check recompilation upon changing header
+  file.create(file_that_exists)
+  with_mocked_cli(compile_ret = list(status = 0), info_ret = list(), code = expect_no_mock_compile({
+    mod$compile(quiet = TRUE, user_header = tmpfile)
```

Comment:
Sorry I'm just not familiar with `testthat`'s mocking schema. Is `mod` here the `mod` from line 36?

```
        mod <- cmdstan_model(
          stan_file = testing_stan_file("bernoulli_external"),
          exe_file = file_that_exists,
          user_header = tmpfile
        )
```

Reply:
um....good question. I don't think so because that shouldn't be in scope anymore. Which begs the question "where is this `mod` object coming from. I'll look into it.

Reply:
lol thank you. Odd!!

Reply:
Looking close, I think you're guess was correct it is getting defined above. `with_mock` puts objects into `topenv()` so parent frame.